### PR TITLE
JustKnobsConfig: Make this work with install_config_module

### PR DESCRIPTION
**Pull request by c00w**
The goal of this PR is to make JustKnobsConfig usable with install_config_module. in support of this, there are 3 major parts to this PR.

- The first is that this is based upon https://github.com/pytorch/pytorch/pull/138377, which was broken out into a seperate PR, since it's mostly unrelated to JK changes.
- The second is rewrite install_config_modules to use a layered approach. Explicitly keeping user_overrides and default values seperate. This allows easy insertion of JK default values as a layer in between (and in the future other values if desired, i.e. env variables).
- The last one is convert JustKnobsConfig to be a pure dataclass, and do all resolution within the existing install_config_module resolution system.

There are a couple design decisions that should be mentioned.
1) I flipped everything to a single map, storing _ConfigEntries. We could use multiple maps but this seemed simpler.
2) When saving config we convert everything (including JKs) into user overrides when the configuration is loaded back in. We could do more complex pickling, but that seemed unlikely to have a lot of use.

let me know how this looks.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

**Review by oulgen — APPROVED** [review]

File: torch/testing/_internal/fake_config_module.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+from torch.utils._config_module import install_config_module
+from torch._utils_internal import JustKnobsConfig
+
+example_knob = JustKnobsConfig(name="true_knob")
+example_knob_false = JustKnobsConfig(name="false_knob", default=False)
+example_knob_force= JustKnobsConfig(name="false_knob", default=False, override=True)
+_cache_config_ignore_prefix = []
+
+install_config_module(sys.modules[__name__])
```
Comment (ezyang) [inline torch/testing/_internal/fake_config_module.py:11]: So are you going to test this module?

**Review by ezyang — CHANGES_REQUESTED** [review]
This seems pretty surprising. It's possible that you have made the right choice, but the PR desc is not enough for me to convinced it should be done this way. And there is not enough exercise of the end-to-end flow for me to evaluate (the test case is too simplified for me to understand)

**Comment by c00w** [comment]
@ezyang my apologies, I forget to add the 2 files that actually matter here (Namely the changes to _config_module and _util_internal). Let me know what questions you have with those 2 files pushed.

File: torch/testing/_internal/fake_config_module.py (hunk 1 above)
Comment (c00w) [reply torch/testing/_internal/fake_config_module.py:11]: Yes - sorry forgot to add the files, please take another look.

File: torch/_utils_internal.py [hunk 2]
```diff
@@ -182,15 +182,17 @@ class JustKnobsConfig:
     foo.bar.set(True)
 
     Note that this does allow for no JK name (so that you can use this to replace old configurations).
+
+    This config will "reset" iself when the config is saved, unless it's been manually overriden, in which case it'll retain the overriden value.
     """
 
     def __init__(
-        self, *, name: Optional[str] = None, env_name=None, default: bool = True
+        self, *, name: Optional[str] = None, env_name=None, default: bool = True, override: Optional[bool] = None
```
Comment (ezyang) [inline torch/_utils_internal.py:190]: It's not immediately obvious to me what override does, so it needs docs

File: torch/utils/_config_module.py [hunk 3]
```diff
@@ -6,15 +6,15 @@
...
+from typing import Any, Callable, Dict, NoReturn, Optional, Set, Union 
 from typing_extensions import deprecated
-from unittest import mock
-
+from unittest import mock 
+from torch._utils_internal import JustKnobsConfig
 
 # Types saved/loaded in configs
-CONFIG_TYPES = (int, float, bool, type(None), str, list, set, tuple, dict)
+CONFIG_TYPES = (int, float, bool, type(None), str, list, set, tuple, dict, JustKnobsConfig)
```
Comment (ezyang) [inline torch/utils/_config_module.py:17]: This is really suspicious

File: test/test_utils_internal.py [hunk 4]
```diff
@@ -13,9 +14,46 @@
...
 class TestJustKnob(TestCase):
+    def test_justknob_config_module(self):
+        with self.subTest("saving+loading works correctly"):
+            fake_config_module.load_config(fake_config_module.save_config())
+        with self.subTest("portable converts to int"):
+            out = fake_config_module.save_config_portable()
+            self.assertEqual(out, {"example_knob_force": True})
+        with self.subTest("codegen config"):
+            out = fake_config_module.codegen_config()
+            self.assertEqual(out, "torch.testing._internal.fake_config_module.example_knob_force = True")
```
Comment (ezyang) [inline test/test_utils_internal.py:29]: There's no dependence between these tests. There's no reason to use subTest. Just make separate test methods for each of them if you really want to tell the difference

File: torch/utils/_config_module.py (hunk 3 above)
Comment (c00w) [reply torch/utils/_config_module.py]: No longer present.

File: test/test_utils_internal.py (hunk 4 above)
Comment (c00w) [reply test/test_utils_internal.py]: Tests gone, but in general removed.

File: test/test_utils_internal.py [hunk 5]
```diff
@@ -13,9 +14,46 @@
 load_tests = load_tests
 
 from torch.testing._internal.common_utils import run_tests, TestCase
+from torch.testing._internal import fake_config_module
 
 
 class TestJustKnob(TestCase):
+    def test_justknob_config_module(self):
+        with self.subTest("saving+loading works correctly"):
+            fake_config_module.load_config(fake_config_module.save_config())
```
[the remark this answers is not in the document]
Comment (c00w) [reply test/test_utils_internal.py]: Test removed, but in general done.

**Comment by c00w** [comment]
@ezyang  - PR is updated per offline discussion, if you can take another look, that would be awesome :).

**Comment by ezyang** [comment]
Use ghstack to stack diffs and have a usable review view here

**Comment by c00w** [comment]
Ack
First new PR - https://github.com/pytorch/pytorch/pull/138758

**Pull request closed without merging**